cmd/streaming-signature-v4.go

```diff
@@ -297,14 +297,24 @@ func (cr *s3ChunkedReader) Read(buf []byte) (n int, err error) {
 		return n, cr.err
 	}
 
-	if cap(cr.buffer) < size {
-		cr.buffer = make([]byte, size)
+	// Choose whether to use internal buffer or use given buffer based on the size.
+	// This will allow us to remove copy call, and improve performance since this
+	// chuckreader is heavily used in minio
+	var chosenBuf []byte
+	internalBuf := cap(buf) < size
```

Comment:
```suggestion
	internalBuf := len(buf) < size
```

We have to check the length of the caller-provided buffer - not its capacity.
The caller will only consume (at most) `len(buf)` bytes.